lib/view_component/compiler.rb

```diff
@@ -13,17 +13,23 @@ def compiled?
     def compile(raise_errors: false)
       return if compiled?
 
-      if template_errors.present?
-        raise ViewComponent::TemplateError.new(template_errors) if raise_errors
-        return false
-      end
+      subclass_instance_methods = component_class.instance_methods(false)
 
-      if component_class.instance_methods(false).include?(:before_render_check)
+      if subclass_instance_methods.include?(:before_render_check)
         ActiveSupport::Deprecation.warn(
           "`before_render_check` will be removed in v3.0.0. Use `before_render` instead."
         )
       end
 
+      if subclass_instance_methods.include?(:with_content) && raise_errors
+        raise ViewComponent::ComponentError.new("#{component_class} implements a reserved method, `with_content`.")
```

Comment:
We currently use `ArgumentError` in `Base#validate_initialization_parameters!`. If we're going to introduce a custom error class, let's use it there too. I'm fine with either approach (ArgumentError or ComponentError)

Reply:
I'd argue they are different kind of errors. `validate_initialization_parameters` checks for **arguments** in the initializer method, so `ArgumentError` makes sense.
Here, I'm checking if a **method** exists. Using `ArgumentError` could be confusing

Reply:
By that logic, I think that `ArgumentError` is incorrect, as I _believe_ it's meant to be used when an argument passed to a method is incorrect, not that the method's argument declaration itself is incorrect. Maybe we just move to `ComponentError` instead for both?